runtime/vm/ObjectMonitor.cpp

```diff
@@ -283,31 +310,119 @@ objectMonitorEnterNonBlocking(J9VMThread *currentThread, j9object_t object)
+					/*
+					 * Clears the Learning bit and sets the Reserved bit to change the state to Reserved.
+					 * Clears the bits under OBJECT_HEADER_LOCK_RECURSION_MASK then fills in the new RC.
+					 * This is done to reclaim the bits used for the Learning Counter that are no longer used under Reserved.
+					 */
+					incremented = (incremented & ~((j9objectmonitor_t)OBJECT_HEADER_LOCK_LEARNING | (j9objectmonitor_t)OBJECT_HEADER_LOCK_RECURSION_MASK)) |
+					              (J9_FLATLOCK_COUNT(incremented) << OBJECT_HEADER_LOCK_V2_RECURSION_OFFSET) |
+					              OBJECT_HEADER_LOCK_RESERVED;
+					reservedTransition = true;
+				} else {
+					/* LC field overflow check. Don't increment if it will overflow the counter. */
+					if ((lock & OBJECT_HEADER_LOCK_LEARNING_LC_MASK) != OBJECT_HEADER_LOCK_LEARNING_LC_MASK) {
+						incremented = incremented + OBJECT_HEADER_LOCK_LEARNING_FIRST_LC_BIT;
+					}
+				}
+
+				if (lock == VM_ObjectMonitor::compareAndSwapLockword(currentThread, lwEA, lock, incremented, false)) {
+					if (0 == J9_FLATLOCK_COUNT(lock)) {
+						VM_AtomicSupport::monitorEnterBarrier();
+					}
+					if (reservedTransition) {
+						/* Transition from Learning to Reserved occurred so the Reserved Counter in the object's J9Class is incremented by 1. */
+						VM_ObjectMonitor::incrementReservedCounter(J9OBJECT_CLAZZ(currentThread, object));
+					}
+				} else {
+					/*
+					 * Another thread has attempted to get the lock and atomically changed the state to Flat.
+					 * Start over and read the new lockword. The lock can not go back to Learning so this will
+					 * only happen once.
+					 */
+					goto restart;
+				}
+			}
 			/* no barrier is required in the recursive case */
 		} else {
 			/* check to see if object is unlocked (JIT did not do initial inline sequence due to insufficient static type info) */
-			if ((0 == (lock & ~OBJECT_HEADER_LOCK_RESERVED)) &&
+			if ((0 == (lock & ~(j9objectmonitor_t)OBJECT_HEADER_LOCK_RESERVED & ~(j9objectmonitor_t)OBJECT_HEADER_LOCK_LEARNING)) &&
 				VM_ObjectMonitor::inlineFastInitAndEnterMonitor(currentThread, lwEA, false, lock)
 			) {
 				/* compare and swap succeeded - barrier already performed */
+				if (lock == OBJECT_HEADER_LOCK_RESERVED) {
+					/* Object in New-AutoReserve was locked so the Reserved Counter in the object's J9Class is incremented by 1. */
+					VM_ObjectMonitor::incrementReservedCounter(J9OBJECT_CLAZZ(currentThread, object));
+				}
 			} else {
 				J9ObjectMonitor *objectMonitor = NULL;
 				/* check if the monitor is flat */
 				if (!J9_LOCK_IS_INFLATED(lock)) {
 					/* lock is flat, am I the owner? */
 					if (J9_FLATLOCK_OWNER(lock) == currentThread) {
+						/* Convert Learning to Flat instead Inflated for dealing with RC overflow.*/
+						if (OBJECT_HEADER_LOCK_LEARNING == (lock & (OBJECT_HEADER_LOCK_LEARNING + OBJECT_HEADER_LOCK_INFLATED))) {
```

Comment:
Bits should be combined with `|`, not `+`.

Reply:
I'll change that.